core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -991,7 +991,7 @@ public Builder assignUUID(String newUuid) {
 
     // it is only safe to set the format version directly while creating tables
     // in all other cases, use upgradeFormatVersion
-    private Builder setInitialFormatVersion(int newFormatVersion) {
+    public Builder setInitialFormatVersion(int newFormatVersion) {
```

Comment:
the changes in this PR LGTM, but we need to be careful about making this method public (see also https://github.com/apache/iceberg/pull/8381#discussion_r1303428449). Let me think a bit about this to see whether there's a slightly different way to approach this

Reply:
one alternative would be to lazily initialize `formatVersion` inside `TableMetadata.Builder` so that calling `setInitialFormatVersion` wouldn't be necessary from `CatalogHandlers` during table creation. Let's see what others think here
/cc @aokolnychyi @rdblue @amogh-jahagirdar @RussellSpitzer

Reply:
This should not be made public.

Reply:
@hantangwangd I've raised a PR to your branch https://github.com/hantangwangd/iceberg/pull/1/files, but missed @nastra on the lazy formatVersion initialization. Not sure entirely what that meant, but in my approach we add a new `buildFromEmpty(int formatVersion` API so that the format version is only initialized once and never mutated (unless going through an actual upgrade API call)